### ISSUE TITLE
Package maintenance

### ISSUE THREAD

**Issue by bingnz** [MinistryOfMagic/xunit.frameworks.autofac#4]
Hi @dabide, there are a couple of really good updates that it would be great to have included in this NuGet package. There's [this one](https://github.com/alanbarros/xunit.frameworks.autofac/pull/1) from @kemmis, and [this update](https://github.com/MinistryOfMagic/xunit.frameworks.autofac/pull/2) from Alan. Also there's another fork by @dennisroche that has [this PR](https://github.com/dennisroche/xunit.ioc.autofac/pull/17).

Are there any plans to merge these things and perhaps to reintegrate all of the forks of this project? The version update looks pretty straightforward simply to merge in, and the issue with exponential theories is a pretty major bug. It'd be a shame to have even more forks and even more NuGet packages as people work around these things.

Thanks!
David

**Comment by dabide** [comment]
@bingnz I am sorry for neglecting the project this long. It has sort of worked where it was used, so I haven't given it much thought. Recently, I noticed that some tests in theories don't run, which caused me to miss an obvious bug. I didn't know of the PRs you mention, as they weren't created against this repo. I will gladly merge them and others.

**Comment by dabide** [comment]
A bit doubtful about the fork by @dennisroche (https://github.com/dennisroche/xunit.ioc.autofac/pull/17). I see the point of using DI, but is it worth the effort to merge it, @bingnz ? The code in this version is substantially different from the original version.

**Comment by dennisroche** [comment]
@dabide I was the original author of this package, which was then forked here. 

I decided not to continue developing the package as don't use `Autofac` anymore for projects and don't see the reason/value of DI into test fixtures.

**Comment by dabide** [comment]
@dennisroche Sorry, I see now that the fork was by @ze1tgeist.

Personally, I think using DI in tests adds enough value to be warranted, especially when writing integration tests, but also when writing unit test (i.e. not having to mock all the constructor parameters for every test case, and thus avoiding the test breaking because of added, unrelated dependencies). There are some dangers, however, but the ones I can think of would be the same when manually mocking dependencies. Anyway, thanks for doing the initial work on this package!

**Comment by dennisroche** [comment]
@dabide No problems 😃 I'm happy to see the idea continue on - I had forgotten to mark the original repository as read-only. This implementation is the preferred fork. 

> using DI in tests adds enough value to be warranted, especially when writing integration tests, but also when writing unit test 

Perhaps we test differently... I prefer testing from the API using the `WebApplicationFactory` in a fixture, using as many real services as possible. Any DI into the fixture is from the test framework (i.e. xUnit). This approach is also flexible to test specific services as you can resolve from the `WebApplicationFactory` via `factory.Services.GetRequiredService()`.

**Comment by bingnz** [comment]
Hi @dabide, thanks for your response!

For others looking at this thread as well, I put some instructions in the proposed README update [here](https://github.com/MinistryOfMagic/xunit.frameworks.autofac/pull/5) that talk about using [this library](https://github.com/pengweiqhca/Xunit.DependencyInjection) instead, as it has equivalent functionality but it's more general (i.e., you can bring in which ever DI container you like). I'm not sure if that has any bearing for you on whether you continue updating this project, but I've found that it was pretty straightforward to switch over all of our projects to use that library instead with minimal changes - and all the problems like having theories exploding out exponentially went away.

@dennisroche I test that way as well; effectively the unit tests isolate and mock only the boundaries of the system where it uses an external resource outside its own control (e.g., something on the Internet). For an API project I'd use TestServer and leave the internal IoC doing whatever it wants to but replace only the abstractions that rely on external APIs or a database, for example. The XUnit IoC libraries such as this one don't preclude that way of testing at all or even discourage it - but they do lead to better-factored tests that might share components like retry policies, data builders, and other test utilities by injecting those things into the tests or fixtures themselves. We expect good tests to remain unchanged for longer than the code they're testing if they're going to guide refactoring - so to me it makes sense that we'd want to use good practices like dependency inversion etc. in test projects too. That's why I personally find IoC useful in XUnit projects; particularly end-to-end test projects but often in unit test projects as well. It can be overkill or lead to confusion sometimes, but in many situations it really leads to clearer tests.

**Comment by dennisroche** [comment]
@bingnz I will give the `Xunit.DependencyInjection` a go. Thanks for the tip. Does that mean you have a second IoC container for just testing concerns?

**Comment by dabide** [comment]
@bingnz Thank's for the tip about `Xunit.DependencyInjection`. For what I've seen so far, it is not feature equivalent (for instance, `ICollectionFixture<T>` and `IClassFixture<T>`), so I think I'll still try to fix this one. :)

**Comment by bingnz** [comment]
@dennisroche no, you can still bring in Autofac over the top since that plugs into the Microsoft abstractions. You could choose to use a different container or just use the Microsoft one though if you wanted, but we use Autofac (basically you add `UseServiceProviderFactory(new AutofacServiceProviderFactory())` in your `ConfigureHost(IHostBuilder hostBuilder)` method in Startup.cs.

@dabide yes it doesn't have the same hooks that this one does - it doesn't support the custom lifetime scopes for instance - but it had feature parity with everything we were using. It'd still be useful to have an Autofac-specific version I'm sure!

**Comment by kemmis** [comment]
@dabide I found the issue with theories running duplicate times. Would you be able to publish a new package if I create a PR? 

I compared your AutofacTheoryDiscoverer to how the regular TheoryDiscoverer works and found that `CreateTestCasesForDataRow()` needs to create a single test case rather than a theory test case. Likewise for `CreateTestCasesForSkip()`. Here is the fix: 

![image](https://user-images.githubusercontent.com/647545/150960649-bb6aee27-1e3b-41e0-bf71-0342301e5bc9.png)

And here is a unit test to verify the number of times theories should execute:

```
private static int _numRuns = 0;

[Theory]
[InlineData("ABC-02-04-CDE")]
[InlineData("ABC-02-40-CDE")]
[InlineData("ABC-02-0040-CDE")]
public void Theories_run_correct_num_times(string input)
{
    var thisTestMethod = MethodInfo.GetCurrentMethod();
    var inlineDataAttributes = thisTestMethod.GetCustomAttributes(typeof(InlineDataAttribute), false);
    var numInlineDataItems = inlineDataAttributes.Length;

    input.Should().Be(input);

    _numRuns++;
    Assert.True(_numRuns <= numInlineDataItems,
                $"Theory should only run {numInlineDataItems} times. It has run {_numRuns} times now.");
}
```

**Comment by dabide** [comment]
@kemmis A pull request would be very welcome!

**Comment by kemmis** [comment]
@dabide here is my PR: https://github.com/MinistryOfMagic/xunit.frameworks.autofac/pull/8

**Comment by dabide** [comment]
Thanks a lot, @kemmis ! Your PR has been merged, and v0.5.1 has been pushed to Nuget.org
